packages/protocol/contracts/layer1/preconf/impl/PreconfSlasher.sol

```diff
@@ -102,69 +102,81 @@ contract PreconfSlasher is IPreconfSlasher, EssentialContract {
+        require(_payload.anchorHash == batchInfo.anchorBlockHash, PossibleReorgOfAnchorBlock());
 
         // Check for reorgs if the committer missed the proposal
-        if (evidence.batchInfo.proposer != _committer) {
+        if (batchInfo.proposer != _committer) {
             // If the beacon block root is not available, it means that the preconfirmed block
             // was reorged out due to an L1 reorg.
             if (LibPreconfUtils.getBeaconBlockRootAt(_payload.preconferSlotTimestamp) == 0) {
                 return getSlashAmount().reorgedPreconf;
             }
         }
 
+        LibBlockHeader.BlockHeader memory actual = evidence.actualBlockHeader;
+        require(actual.number == blockId, InvalidActualBlockHeader());
         // The preconfirmed blockhash must not match the hash of the proposed block for a
         // preconfirmation violation
-        bytes32 actualBlockHash = evidence.blockhashProofs.actualBlockHeader.hash();
+        bytes32 actualBlockHash = actual.hash();
         require(_payload.blockHash != actualBlockHash, PreconfirmationIsValid());
 
+        LibBlockHeader.BlockHeader memory verified = evidence.verifiedBlockHeader;
+        require(verified.number == batchInfo.lastBlockId, InvalidVerifiedBlockHeader());
+
         // Validate that the batch has been verified
         ITaikoInbox.TransitionState memory transition =
             taikoInbox.v4GetBatchVerifyingTransition(uint64(_payload.batchId));
 
         // Validate the verified blockheader
-        require(
-            transition.blockHash == evidence.verifiedBlockHeader.hash(),
-            InvalidVerifiedBlockHeader()
-        );
-
-        uint256 blockId = evidence.preconfedBlockHeader.number;
+        require(transition.blockHash == verified.hash(), InvalidVerifiedBlockHeader());
 
         // Validate that the parent on which this block was preconfirmed made it to the inbox, i.e
         // the parentHash within the preconfirmed block header must match the hash of the proposed
         // parent.
         LibTrieProof.verifyMerkleProof(
-            evidence.verifiedBlockHeader.stateRoot,
+            verified.stateRoot,
             taikoAnchor,
             _calcBlockHashSlot(blockId - 1),
-            evidence.preconfedBlockHeader.parentHash,
+            preconfed.parentHash,
             evidence.parentBlockhashProofs.accountProof,
             evidence.parentBlockhashProofs.storageProof
         );
 
-        // Verify that `blockhashProofs` correctly proves the blockhash of the block proposed
-        // at the same height as the preconfirmed block.
-        LibTrieProof.verifyMerkleProof(
-            evidence.verifiedBlockHeader.stateRoot,
-            taikoAnchor,
-            _calcBlockHashSlot(blockId),
-            actualBlockHash,
-            evidence.blockhashProofs.accountProof,
-            evidence.blockhashProofs.storageProof
-        );
+        if (verified.number == blockId) {
```

Comment:
if  the block in question is the last block in a batch, even if this batch is verified, it block's hash is not written into storage by anchor tx -- anchor only write the parent block hash, not the current block hash into storage.